torchserve/datadog_checks/torchserve/model_discovery.py

```diff
@@ -0,0 +1,67 @@
+# (C) Datadog, Inc. 2023-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from datadog_checks.base import AgentCheck
+from datadog_checks.base.checks.kubelet_base.base import urljoin
+from datadog_checks.base.utils.discovery import Discovery
+
+
+class ModelDiscovery(Discovery):
+    def __init__(self, check, limit=None, include=None, exclude=None, interval=None):
+        super().__init__(
+            self.get_models,
+            limit=limit,
+            include={pattern: None for pattern in include} if include else None,
+            exclude=exclude,
+            interval=interval,
+            key=lambda n: n.get("modelName"),
+        )
+        self.check = check
+        self.api_status = AgentCheck.UNKNOWN
+        self.all_models = None
+        # Used to know when the cache used internally in `Discovery` gets refreshed.
+        # Useful to avoid trying to parse the payloads to generate events in case it's not needed
+        self.refreshed = False
+
+    def get_items(self):
+        # Force the flag to False to make sure we do not keep the previous value.
+        self.refreshed = False
+        return super().get_items()
+
+    def get_models(self):
```

Comment:
This could be simplified with recursion

```
def get_models(self, page_number:int = 0, models: list = [], params: dict = {"limit": self.check.instance["limit"]}) -> list:
        try:
            url = urljoin(self.check.instance["management_api_url"], "models")
            self.check.log.debug(f"Querying page {page_number}")
            response = self.check.http.get(url, params=params)
            response.raise_for_status()
            self.api_status = AgentCheck.OK

            content = response.json()
            models.extend(content.get("models", []))

            if "nextPageToken" in content:
                self.get_models(
                       page_number = page_number + 1,
                       models = models,
                       params = params | {"nextPageToken": content["nextPageToken"]} # Supported in python3.9+
                )
        except Exception as e:
            self.check.log.error(f"Caught exception {e}, no models for page {page_number} with token {content["nextPageToken"]}")
            self.api_status = AgentCheck.CRITICAL if page_number == 0 else self.api_status
        else:
            # The cache wanted to refresh, so we update the flag
            self.refreshed = True
            self.check.gauge("models", len(models), tags=self.check.tags)
            self.all_models = {model["modelName"]: model["modelUrl"] for model in models}

        return models
```

Reply:
Not sure simplified is the right word. Also you are emitted several times `self.check.gauge("models", len(models), tags=self.check.tags)` here

Reply:
Yes, it should be once similar to api status update. Good catch 👍 

`self.check.gauge("models", len(models), tags=self.check.tags) if page_number == 0 else None`

I have tendency to use recursion, so I might be biased for recursion with the simplicity comment 😄 The functionality would remain the same in both cases, so I will leave it for you to decide iterative/recursive approach 😉

Reply:
It's a matter of taste I guess! I'll stick with the iterative approach here